### ISSUE TITLE
Deadlock detected

### ISSUE THREAD

**Issue by schneems** [ryansmith3136/lpxc#7]
Have you seen this before by any chance? The error is very intermittent. We're vendoring in the latest lpxc into the buildpack:

https://discussion.heroku.com/t/lock-deadlock-detected-fatal/271/3

```
<internal:prelude>:8:in `lock': deadlock detected (fatal)
    from <internal:prelude>:8:in `synchronize'
    from /app/tmp/buildpacks/ruby/vendor/lpxc.rb:57:in `puts'
    from /app/tmp/buildpacks/ruby/lib/language_pack/instrument.rb:10:in `bench_msg'
    from /app/tmp/buildpacks/ruby/lib/language_pack/instrument.rb:23:in `instrument'
    from /app/tmp/buildpacks/ruby/lib/language_pack/base.rb:43:in `instrument'
    from /app/tmp/buildpacks/ruby/lib/language_pack/base.rb:39:in `instrument'
    from /app/tmp/buildpacks/ruby/lib/language_pack/rails4.rb:72:in `run_assets_precompile_rake_task'
    from /app/tmp/buildpacks/ruby/lib/language_pack/ruby.rb:121:in `block (2 levels) in compile'
    from /app/tmp/buildpacks/ruby/lib/language_pack/ruby.rb:668:in `call'
    from /app/tmp/buildpacks/ruby/lib/language_pack/ruby.rb:668:in `allow_git'
    from /app/tmp/buildpacks/ruby/lib/language_pack/ruby.rb:116:in `block in compile'
    from /app/tmp/buildpacks/ruby/lib/language_pack/instrument.rb:19:in `block (2 levels) in instrument'
    from /app/tmp/buildpacks/ruby/lib/language_pack/instrument.rb:41:in `yield_with_block_depth'
    from /app/tmp/buildpacks/ruby/lib/language_pack/instrument.rb:18:in `block in instrument'
    from /usr/local/lib/ruby/1.9.1/benchmark.rb:310:in `realtime'
```

I know there was a deadlock issue earlier that was resolved, could this be releated, or perhaps something new?

**Comment by ryansmith3136** [comment]
@schneems Sorry for the delayed response. Any updates on your end regarding this issue?

This is a 1 in a million bug from what I can tell. I wish I had a core dump so I can do a proper post-mortem. It is hard for me to understand why this happens. I am going to re-read the locking portions of the library to see if I can spot anything obvious.

If you can, please help review the locking portion of the code as well. The library is fairly simple and it shouldn't take you much effort to understand what is going on. Let me know if I can help aid in understanding.

**Comment by ryansmith3136** [comment]
I welcome anyone to help understand where lxpc might deadlock. I suspect there is a great learning opportunity in providing a reason as to why this deadlock occurred. Unfortunately, we don't have a lot of evidence or post-mortem details to aid in debugging. I would first suggest writing a driver that will surface the deadlock. Another approach could be studying the code in hopes of pointing out an obvious flaw.

**Comment by glenngillen** [comment]
@schneems do you still get a deadlock if you change [this line](https://github.com/heroku/heroku-buildpack-ruby/blob/master/lib/language_pack/instrument.rb#L29) to:

``` ruby
Thread.current[:out] ||= ENV['LOGPLEX_DEFAULT_TOKEN'] ? Lpxc.new(:disable_delay_flush => true) : StringIO.new
```

?

**Comment by danp** [comment]
Came across [this](http://bugs.ruby-lang.org/issues/show/4009) but looks like it was included in 1.9.2-p290 which is the version `/usr/local/bin/ruby -v` reports in a dyno.

**Comment by schneems** [comment]
Glenn, i'm manually flushing after each line, I could try that though.

**Comment by ryansmith3136** [comment]
@schneems Interesting. I didn't realize that the build-pack was calling `flush()`. Did you know that `flush` is a private method? :trollface:

I need to take a closer look at the code to understand if there exists a scenario by which manually calling flush will cause a deadlock. You should note that [I ask callers of `flush` to lock the mutex](https://github.com/ryandotsmith/lpxc/blob/master/lib/lpxc.rb#L81) on the hash. This locking requirement was one of the reasons I made that method private.

If you don't want messages to be made into batches (1 log message per HTTP request) you can set the batch size to 1. This will produce the same behavior that you are getting now by manually calling flush with the added benefit that you will not be accessing a private method. For example:

``` ruby
lpxc = Lpxc.new(batch_size: 1)
```

**Comment by glenngillen** [comment]
@schneems so I saw. `Lpxc#puts` and `Lpxc#delay_flush` will both try to obtain a lock. Your `Instrument#bench_msg` will try and flush without taking a lock. If you're able to reproduce somewhat reliably, and can then solve it with that extra option it'll help explain whether the issue is a race condition within `Lpxc` or whether it's because you're calling a private API.

**Comment by ryansmith3136** [comment]
Since the build-pack is not disabling the delayed flush and it is calling the flush method then the behavior of lxpc can not be safe. In the aforementioned scenario, when the build-pack is calling flush _without_ using the `@hash_lock` it is possible that many threads are operating on `@hash` simultaneously.

**Comment by ryansmith3136** [comment]
I have updated the test cases so that they do not rely on the flush method. Hopefully this should demonstrate that the flush method is not to be used by client code.

**Comment by schneems** [comment]
I was treating lxpc as a drop in replacement for StringIO which uses flush to clear the buffer. Really it's my fault for not respecting the private and understanding the options better.

**Comment by ryansmith3136** [comment]
@schneems I didn't realize that String.io promoted the use of the flush method. It might be worth making `Lpxc#flush` safe to use so that it blends well with other related objects. Thoughts?

**Comment by schneems** [comment]
The docs for StringIO are pretty sparse and really don't recommend any use: 

http://www.ruby-doc.org/stdlib-2.0/libdoc/stringio/rdoc/StringIO.html
I think we could also set sync to true but flush seemed to be a bit more explicit.

It would be nice to match that interface or to recommend a way to test in the absence of a working logoplex instance. I typically use StringIO when capturing STDOUT in tests so thats why I picked the same class for testing my lxpc code. Just happened both had a flush method.

:+1: Making flush public and threadsafe.

**Comment by ryansmith3136** [comment]
I would love a code review on #8 

Once I get an extra set of eyes on my changes, I will gladly merge and cut a new release.

**Comment by ryansmith3136** [comment]
fixed in #8

**Comment by ryansmith3136** [comment]
Publicly available in [v0.0.6](https://github.com/ryandotsmith/lpxc/releases/tag/v0.0.6)
